grype/pkg/syft_json_provider.go

```diff
@@ -20,33 +21,57 @@ import (
 // resilient to multiple syft JSON schemas (to a degree).
 // TODO: add version detection and multi-parser support (when needed in the future)
 func syftJSONProvider(config providerConfig) ([]Package, Context, error) {
-	var reader io.Reader
+	reader, err := getSyftJSON(config)
+	if err != nil {
+		return nil, Context{}, err
+	}
+
+	return parseSyftJSON(reader)
+}
+
+func getSyftJSON(config providerConfig) (io.Reader, error) {
 	if config.reader != nil {
-		// the caller is explicitly hinting to use the given reader as input
-		reader = config.reader
-	} else {
-		// try and get a reader from the description
-		if strings.HasPrefix(config.userInput, "sbom:") {
-			// the user has explicitly hinted this is an sbom, if there is an issue return the error
-			filepath := strings.TrimPrefix(config.userInput, "sbom:")
-			sbomReader, err := os.Open(filepath)
-			if err != nil {
-				return nil, Context{}, fmt.Errorf("user hinted 'sbom:' but couldn't read SBOM file: %w", err)
-			}
-			reader = sbomReader
-		}
+		// the caller has explicitly indicated to use the given reader as input
+		return config.reader, nil
+	}
 
-		// the user has not hinted that this may be a sbom, but lets try that first...
-		if sbomReader, err := os.Open(config.userInput); err != nil {
-			if errors.Is(err, os.ErrNotExist) {
-				return nil, Context{}, errDoesNotProvide
-			}
-		} else {
-			reader = sbomReader
+	if explicitlySpecifyingSBOM(config.userInput) {
+		filepath := strings.TrimPrefix(config.userInput, "sbom:")
+
+		sbom, err := openSbom(filepath)
+		if err != nil {
+			return nil, fmt.Errorf("unable to use specified SBOM: %w", err)
 		}
+
+		return sbom, nil
 	}
 
-	return parseSyftJSON(reader)
+	// as a last resort, see if the raw user input specified an SBOM file
+	sbom, err := openSbom(config.userInput)
+	if err == nil {
+		return sbom, nil
+	}
+
+	// no usable SBOM is available
+	return nil, errDoesNotProvide
+}
+
+func openSbom(path string) (*os.File, error) {
+	expandedPath, err := homedir.Expand(path)
+	if err != nil {
+		return nil, fmt.Errorf("unable to open SBOM: %w", err)
```

Comment:
I think this error should be specific to expanding the path, not to open the SBOM

Reply:
I can make that change. 👍 

This is one area I think it'd be good to seek alignment on among the group. Errors are interesting... (especially in Go). If function A calls function B, and thus has to handle getting an error back from function B, which function reports that there was a problem with function B's responsibility? Lately, I've been preferring function B to do this, since that consolidates the message to function B, and any number of callers of function B can rely on this error message rather than have them all say something like "there was a problem with function B" in their own function bodies.

In this specific case, if you look into the code for the function `homedir.Expand(path)`, you'll notice it returns errors like `"cannot expand user-specific home dir"`. So in my function, `openSbom`, rather than `fmt.Errorf("error expanding the home dir in the path": %w", err)`, which would produce:

```
ERROR: error expanding the home dir in the path: cannot expand user-specific home dir
```

I decided to have _my_ error reflect the failure in _my function's_ job, which is to open an SBOM.

Again, happy to change it, but just wanted to share those thoughts!

Reply:
That explanation makes sense to me, and I agree with all your points. The caller of this function mentions something similar which seemed repetitive:

```go
    return nil, fmt.Errorf("unable to use specified SBOM: %w", err)
```

Re-using a bit your example, that would mean it would produce this error:

```
ERROR: unable to use specified SBOM: unable to open SBOM: cannot expand user-specific home dir
```

This reads somewhat repetitive. I don't have a good suggestion or alternative here, but that repetition is what made me think that perhaps the messaging should change. 

However! I'm fine with whatever you decide, this is not a hard comment :)

Reply:
That's an interesting point. I'll think about that. I think my thinking was that there could be multiple reasons why an SBOM is unusable (e.g. invalid JSON syntax). `getSyftJSON` doesn't care why the SBOM is unusable, it just needs to know that it's not usable. The inner error `unable to open SBOM...` in `openSbom` is a specific reason why the SBOM was unusable — specifically that the SBOM couldn't be opened.